pkg/pillar/cmd/zedagent/localinfo.go

```diff
@@ -192,81 +203,199 @@ func processReceivedAppCommands(ctx *getconfigContext, cmdList *profile.LocalApp
 			continue
 		}
 		// Try to find the application instance.
-		ais := findAppInstance(ctx, appUUID, displayName)
-		if ais == nil {
+		appInst := findAppInstance(ctx, appUUID, displayName)
+		if appInst == nil {
 			log.Warnf("Failed to find app instance with UUID=%s, displayName=%s",
 				appUUID, displayName)
 			continue
 		}
-		appUUID = ais.UUIDandVersion.UUID
+		appUUID = appInst.UUIDandVersion.UUID
+		if _, duplicate := publishApps[appUUID.String()]; duplicate {
+			log.Warnf("Multiple commands requested for app instance with UUID=%s",
+				appUUID)
+			continue
+		}
+
 		command := types.AppCommand(appCmdReq.Command)
-		appCmd := ctx.localAppCommands.LookupByAppUUID(appUUID)
-		if appCmd != nil {
-			// Entry for this app already exists.
-			if appCmd.Command == command &&
-				appCmd.LocalServerTimestamp == appCmdReq.Timestamp {
-				// already accepted
-				continue
+		appCmd, hasLocalCmd := ctx.localCommands.AppCommands[appUUID.String()]
+		if !hasLocalCmd {
+			appCmd = &types.LocalAppCommand{}
+			if ctx.localCommands.AppCommands == nil {
+				ctx.localCommands.AppCommands = make(map[string]*types.LocalAppCommand)
 			}
-			appCmd.Command = command
-			appCmd.LocalServerTimestamp = appCmdReq.Timestamp
-			appCmd.DeviceTimestamp = time.Now()
-			appCmd.Completed = false
-			cmdChanges = true
+			ctx.localCommands.AppCommands[appUUID.String()] = appCmd
+		}
+		if appCmd.Command == command &&
+			appCmd.LocalServerTimestamp == appCmdReq.Timestamp {
+			// already accepted
 			continue
 		}
-		// Add new entry.
-		ctx.localAppCommands.Cmds = append(ctx.localAppCommands.Cmds,
-			types.LocalAppCommand{
-				AppUUID:              appUUID,
-				Command:              command,
-				LocalServerTimestamp: appCmdReq.Timestamp,
-				DeviceTimestamp:      time.Now(),
-				Completed:            false,
-			})
+		appCmd.Command = command
+		appCmd.LocalServerTimestamp = appCmdReq.Timestamp
+		appCmd.DeviceTimestamp = time.Now()
+		appCmd.Completed = false
+
+		appCounters, hasCounters := ctx.localCommands.AppCounters[appUUID.String()]
+		if !hasCounters {
+			appCounters = &types.LocalAppCounters{}
+			if ctx.localCommands.AppCounters == nil {
+				ctx.localCommands.AppCounters = make(map[string]*types.LocalAppCounters)
+			}
+			ctx.localCommands.AppCounters[appUUID.String()] = appCounters
+		}
+
+		// Update configuration to trigger the operation.
+		switch appCmd.Command {
+		case types.AppCommandRestart:
+			appCounters.RestartCmd.Counter++
+			appCounters.RestartCmd.ApplyTime = appCmd.DeviceTimestamp.String()
+			appInst.LocalRestartCmd = appCounters.RestartCmd
+			publishApps[appUUID.String()] = *appInst
+
+		case types.AppCommandPurge:
+			appCounters.PurgeCmd.Counter++
+			appCounters.PurgeCmd.ApplyTime = appCmd.DeviceTimestamp.String()
+			appInst.LocalPurgeCmd = appCounters.PurgeCmd
```

Comment:
I would add a comment here (or maybe a comment block above) that we Get the AppInstanceConfig and update it wih a new generation counter (and we also do that for the VolumeConfig/VolumeRefConfig for the app instance) and then re-publish those.

Reply:
Done - added separate explanation for Restart and for Purge.